lsp/nls/src/world.rs

```diff
@@ -412,4 +666,283 @@ impl World {
+    ///
+    /// This method spins up a new cache, and pre-fill the old AST representation for the stdlib,
+    /// the given file and the transitive closure of its imports.
+    pub fn cache_hub_for_eval(&self, file_id: FileId) -> CacheHub {
+        use nickel_lang_core::{
+            bytecode::ast::compat::ToMainline,
+            cache::{EntryState, TermEntry},
+        };
+
+        let mut cache = CacheHub::new();
+        cache.sources = self.sources.clone();
+        cache.import_data = self.import_data.clone();
+
+        cache.terms.insert(
+            self.analysis_reg.stdlib_analysis().file_id(),
+            TermEntry {
+                term: self.compiled_stdlib.clone(),
+                state: EntryState::Typechecked,
+                format: InputFormat::Nickel,
+            },
+        );
+
+        cache.terms.insert(
+            file_id,
+            TermEntry {
+                term: self.analysis_reg.get(file_id).unwrap().ast().to_mainline(),
+                state: EntryState::Parsed,
+                format: InputFormat::Nickel,
+            },
+        );
+
+        let mut work_stack: Vec<FileId> = self
+            .import_data
+            .imports
+            .get(&file_id)
+            .map(|imports| imports.iter().copied().collect())
+            .unwrap_or_default();
+
+        // Imports can be cyclic, so we need to keep track of what we've already done.
+        let mut visited = HashSet::new();
+        visited.insert(file_id);
+
+        while let Some(next) = work_stack.pop() {
+            if !visited.insert(next) {
+                continue;
+            }
+
+            // It's fine to not pre-fill the cache: we could start from an empty cache and things
+            // would still work out. We just try to fill as many things as we can to avoid
+            // duplicating work.
+            let Some(analysis) = self.analysis_reg.get(next) else {
+                continue;
+            };
+
+            cache.terms.insert(
+                next,
+                TermEntry {
+                    term: analysis.ast().to_mainline(),
+                    state: EntryState::Typechecked,
+                    format: InputFormat::Nickel,
+                },
+            );
+
+            work_stack.extend(
+                self.import_data
+                    .imports
+                    .get(&next)
+                    .cloned()
+                    .unwrap_or_default(),
+            );
+        }
+
+        cache
+    }
+
+    pub fn get_import_target(&self, pos: TermPos) -> Option<FileId> {
+        let pos = pos.into_opt()?;
+        self.import_targets.get(&pos.src_id)?.get(&pos).copied()
+    }
+}
+
+/// The import resolver used by [World]. It borrows from the analysis registry, from the source
+/// cache and from the import data that are updated as new file are parsed.
+///
+/// TODO: why do we use `new_imports`
```

Comment:
It seems to just be somewhere to store the new `PackedAnalysis`s until they can be inserted in the analysis registry?

Reply:
Ah, I think this is a leftover from when I copy pasted it from the cache, where we also used to have `new_imports` but I got rid of it in the end. Will remove the comment here, since indeed we do use it.